File: cloudstate-operator/pkg/stores/store.go
/*
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package stores

import (
	"context"

	cloudstate "github.com/cloudstateio/cloudstate/cloudstate-operator/pkg/apis/v1alpha1"
	appsv1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	"sigs.k8s.io/controller-runtime/pkg/builder"
)

type Store interface {
	InjectPodStoreConfig(ctx context.Context, name string, namespace string, pod *corev1.Pod, container *corev1.Container, store *cloudstate.StatefulStore) error

	ReconcileStatefulService(ctx context.Context, service *cloudstate.StatefulService, deployment *appsv1.Deployment, store *cloudstate.StatefulStore) ([]cloudstate.CloudstateCondition, error)
	SetupWithStatefulServiceController(builder *builder.Builder) error

	ReconcileStatefulStore(ctx context.Context, store *cloudstate.StatefulStore) (conditions []cloudstate.CloudstateCondition, updated bool, err error)
	SetupWithStatefulStoreController(builder *builder.Builder) error
}

func appendEnvVar(container *corev1.Container, name string, value string) {
	container.Env = append(container.Env, corev1.EnvVar{
		Name:  name,
		Value: value,
	})
}

func appendSecretEnvVar(container *corev1.Container, name string, secret corev1.LocalObjectReference, key string) {
	container.Env = append(container.Env, corev1.EnvVar{
		Name: name,
		ValueFrom: &corev1.EnvVarSource{
			SecretKeyRef: &corev1.SecretKeySelector{
				LocalObjectReference: secret,
				Key:                  key,
			},
		},
	})
}

func appendOptionalSecretEnvVar(container *corev1.Container, name string, secret corev1.LocalObjectReference, key string) {
	true := true
	container.Env = append(container.Env, corev1.EnvVar{
		Name: name,
		ValueFrom: &corev1.EnvVarSource{
			SecretKeyRef: &corev1.SecretKeySelector{
				LocalObjectReference: secret,
				Key:                  key,
				Optional:             &true,
			},
		},
	})
}

File: proxy/core/src/main/scala/io/cloudstate/proxy/valueentity/store/Store.scala
/*
 * Copyright 2019 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.cloudstate.proxy.valueentity.store

import akka.util.ByteString

import scala.concurrent.Future

object Store {

  case class Key(persistentId: String, entityId: String)

  /** Value to persist with its type url and its content */
  case class Value(typeUrl: String, state: ByteString)

}

/**
 * A persistent store for value-based entities.
 */
trait Store {

  /**
   * Retrieve the data for the given key.
   *
   * @param key to retrieve data for
   * @return Some(data) if data exists for the key and None otherwise
   */
  def get(key: Store.Key): Future[Option[Store.Value]]

  /**
   * Insert the data with the given key if it not already exists.
   * Update the data at the given key if it already exists.
   *
   * @param key  to insert or update the entity
   * @param value that should be persisted
   */
  def update(key: Store.Key, value: Store.Value): Future[Unit]

  /**
   * Delete the data for the given key.
   *
   * @param key to delete data.
   */
  def delete(key: Store.Key): Future[Unit]

}
